datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,1739 @@
+            )?;
+            let new_right = adjust_input_keys_down_recursively(
+                right.clone(),
+                join_key_pairs.right_keys,
+            )?;
+            Ok(Arc::new(SortMergeJoinExec::try_new(
+                new_left,
+                new_right,
+                on.clone(),
+                *join_type,
+                sort_options.clone(),
+                *null_equals_null,
+            )?))
+        }
+    } else if let Some(AggregateExec {
+        mode,
+        group_by,
+        aggr_expr,
+        input,
+        input_schema,
+        ..
+    }) = plan_any.downcast_ref::<AggregateExec>()
+    {
+        if parent_required.is_empty() {
+            plan.map_children(|plan| adjust_input_keys_down_recursively(plan, vec![]))
+        } else {
+            match mode {
+                AggregateMode::Final => plan.map_children(|plan| {
+                    adjust_input_keys_down_recursively(plan, vec![])
+                }),
+                AggregateMode::FinalPartitioned | AggregateMode::Partial => {
+                    let out_put_columns = group_by
+                        .expr()
+                        .iter()
+                        .enumerate()
+                        .map(|(index, (_col, name))| Column::new(name, index))
+                        .collect::<Vec<_>>();
+
+                    let out_put_exprs = out_put_columns
+                        .iter()
+                        .map(|c| Arc::new(c.clone()) as Arc<dyn PhysicalExpr>)
+                        .collect::<Vec<_>>();
+
+                    // Check whether the requirements can be satisfied by the Aggregation
+                    if parent_required.len() != out_put_exprs.len()
+                        || expr_list_eq_strict_order(&out_put_exprs, &parent_required)
+                        || !group_by.null_expr().is_empty()
+                    {
+                        plan.map_children(|plan| {
+                            adjust_input_keys_down_recursively(plan, vec![])
+                        })
+                    } else {
+                        let new_positions =
+                            expected_expr_positions(&out_put_exprs, &parent_required);
+                        match new_positions {
+                            Some(positions) => {
+                                let mut new_group_exprs = vec![];
+                                for idx in positions.into_iter() {
+                                    new_group_exprs.push(group_by.expr()[idx].clone());
+                                }
+                                let new_group_by =
+                                    PhysicalGroupBy::new_single(new_group_exprs);
+                                match mode {
+                                    AggregateMode::FinalPartitioned => {
+                                        let new_input =
+                                            adjust_input_keys_down_recursively(
```

Comment:
Since the input of `FinalPartitioned` should be the `Partial` and they should share the same column order, it's safe to call `adjust_input_keys_down_recursively` here.